### DISCUSSION TITLE
optimization airfoil with Cm constraint

### DISCUSSION THREAD

**Discussion by zhuwenguo** [mdolab/dafoam#402] [General]
I am running an airfoil optimization with a Cm constraint which has a nose up (negative CMZ in the tutorial) pitch moment, the result I got is weird( maximize L/D and minimize Cd got similar shape), I am starting from a reflexed airfoil HS520, with an Cmz constraint:

>   self.add_constraint("cruise.aero_post.CMZ", lower=-0.01, upper=0.0, `scaler=1.0)`

It seems I should constrain the curvature of the camber line or miss some other constraint, Does anyone give me some tips? Thanks.
![optimization](https://user-images.githubusercontent.com/53405142/221877023-c33a14d1-e3dc-4100-b9ff-5f9e94fb07fd.png)

**Comment by friedenhe** [comment]
Have you tried to run the optimization without the Cm constraint?

**Reply by zhuwenguo** [reply]
yes, the optimization(max L/D) result without the Cm constraint is quite reasonable. 

![without_Cm_constraint](https://user-images.githubusercontent.com/53405142/223114267-2bfeee02-2ec9-4280-9cee-9592c7516157.png)

**Comment by friedenhe** [comment]
Which optimizer are you using? Check the optimality and feasibility of the CM-constrained case. It is possible that that case did not converge well

**Reply by zhuwenguo** [reply]
Sorry to respond so late. I am using the default optimizer IPOPT.

`parser.add_argument("-optimizer", help="optimizer to use", type=str, default="IPOPT")`

Since it reported that "Primal solution failed” error  I have changed the primal res to 1e-3.
`daOptions = {
    "designSurfaces": ["wing"],
    "solverName": "DASimpleFoam",
    "primalMinResTol": 1.0e-3,
    "primalBC": {
        "U0": {"variable": "U", "patches": ["inout"], "value": [U0, 0.0, 0.0]},
        "p0": {"variable": "p", "patches": ["inout"], "value": [p0]},
        "nuTilda0": {"variable": "nuTilda", "patches": ["inout"], "value": [nuTilda0]},
        "useWallFunction": True,
    },`

**Reply by zhuwenguo** [reply]
It is true the optimization does not converge. The opt_IPOPT file is as follows.

> List of user-set options:
> 
>                                     Name   Value                used
>                              alpha_for_y = full                  yes
>                          constr_viol_tol = 1e-05                 yes
>                         file_print_level = 5                     yes
>                    hessian_approximation = limited-memory        yes
>               limited_memory_max_history = 10                    yes
>                            linear_solver = mumps                 yes
>                                 max_iter = 100                   yes
>                              mu_strategy = adaptive              yes
>                       nlp_scaling_method = none                  yes
>                              output_file = opt_IPOPT.txt         yes
>                              print_level = 5                     yes
>                       print_user_options = yes                   yes
>                                 recalc_y = yes                   yes
>                                       sb = yes                   yes
>                                      tol = 1e-05                 yes
> This is Ipopt version 3.13.5, running with linear solver mumps.
> NOTE: Other linear solvers might be more efficient (see Ipopt documentation).
> 
> Number of nonzeros in equality constraint Jacobian...:      208
> Number of nonzeros in inequality constraint Jacobian.:      420
> Number of nonzeros in Lagrangian Hessian.............:        0
> 
> Total number of variables............................:       20
>                      variables with only lower bounds:        0
>                 variables with lower and upper bounds:       20
>                      variables with only upper bounds:        0
> Total number of equality constraints.................:       14
> Total number of inequality constraints...............:       21
>         inequality constraints with only lower bounds:        1
>    inequality constraints with lower and upper bounds:       20
>         inequality constraints with only upper bounds:        0
> 
> iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
>    0 -3.3383357e+01 0.00e+00 1.02e+02   0.0 0.00e+00    -  0.00e+00 0.00e+00   0
> Warning: Cutting back alpha due to evaluation error
> Warning: Cutting back alpha due to evaluation error
> Warning: Cutting back alpha due to evaluation error
>    1 -3.3299324e+01 0.00e+00 4.79e+02   1.5 3.61e+01    -  1.40e-01 3.43e-03f  4
>    2 -4.2770625e+01 0.00e+00 2.26e+01  -0.7 2.16e-02    -  9.95e-01 1.00e+00f  1
>    3 -6.2401170e+01 0.00e+00 1.98e+01  -1.8 5.89e-02    -  1.00e+00 1.00e+00f  1
>    4 -7.1306508e+01 0.00e+00 1.28e+01  -1.5 4.95e+00    -  1.00e+00 1.58e-02f  1
>    5 -4.1974838e+01 0.00e+00 2.64e+01  -2.6 9.75e-03    -  1.00e+00 8.88e-16h 51
>    6 -4.4808673e+01 0.00e+00 3.62e+00  -2.7 2.80e-02    -  8.58e-01 1.00e+00f  1
>    7 -4.5464683e+01 0.00e+00 1.66e+00  -2.7 1.88e-02    -  1.00e+00 5.71e-02f  1
>    8 -6.0528519e+01 0.00e+00 4.44e+01  -2.7 4.18e-03    -  1.00e+00 1.00e+00f  1
> Restoration phase is called at point that is almost feasible,
>   with constraint violation 6.106227e-15. Abort.
> 
> Number of Iterations....: 8
> 
>                                    (scaled)                 (unscaled)
> Objective...............:  -6.0528519254927872e+01   -6.0528519254927872e+01
> Dual infeasibility......:   4.4375775724368516e+01    4.4375775724368516e+01
> Constraint violation....:   0.0000000000000000e+00    0.0000000000000000e+00
> Complementarity.........:   1.9357044482705237e-03    1.9357044482705237e-03
> Overall NLP error.......:   4.4375775724368516e+01    4.4375775724368516e+01
> 
> 
> Number of objective function evaluations             = 137
> Number of objective gradient evaluations             = 10
> Number of equality constraint evaluations            = 140
> Number of inequality constraint evaluations          = 137
> Number of equality constraint Jacobian evaluations   = 10
> Number of inequality constraint Jacobian evaluations = 10
> Number of Lagrangian Hessian evaluations             = 0
> Total CPU secs in IPOPT (w/o function evaluations)   =      0.015
> Total CPU secs in NLP function evaluations           =     76.763
> 
> EXIT: Restoration Failed!

**Reply by friedenhe** [reply]
> Sorry to respond so late. I am using the default optimizer IPOPT.
> 
> `parser.add_argument("-optimizer", help="optimizer to use", type=str, default="IPOPT")`
> 
> Since it reported that "Primal solution failed” error I have changed the primal res to 1e-3. `daOptions = { "designSurfaces": ["wing"], "solverName": "DASimpleFoam", "primalMinResTol": 1.0e-3, "primalBC": { "U0": {"variable": "U", "patches": ["inout"], "value": [U0, 0.0, 0.0]}, "p0": {"variable": "p", "patches": ["inout"], "value": [p0]}, "nuTilda0": {"variable": "nuTilda", "patches": ["inout"], "value": [nuTilda0]}, "useWallFunction": True, },`

If the flow (primal) does not converge well (like ~3 orders), the optimization will probably have problems. I suggest you use primalMinResTol=1e-6.

**Reply by zhuwenguo** [reply]
Hi, Prof.He, if I set the primalMinResTol=1e-6., it will report the error "pyDAFoam Error: Primal solution failed for the baseline design! "

**Comment by friedenhe** [comment]
Also, it seems the log you sent says the problem is unconstrained, inf_pr=0. Is the log file for the constrained Cm case? Another thing you can do is to use a different optimizer, like SLSQP or SNOPT.

**Reply by zhuwenguo** [reply]
Prof.He, here is the files I used to optimization. I have added the constraints:
```
# add objective and constraints to the top level
self.add_objective("LoD.val", scaler=-1.0)
```
`self.connect("cruise.aero_post.CD", "LoD.CD")`
`self.connect("cruise.aero_post.CL", "LoD.CL")`
`self.add_constraint("geometry.thickcon", lower=0.5, upper=3.0, scaler=1.0)`
`self.add_constraint("geometry.volcon", lower=1.0, scaler=1.0)`
`self.add_constraint("geometry.tecon", equals=0.0, scaler=1.0, linear=True)`
`self.add_constraint("geometry.lecon", equals=0.0, scaler=1.0, linear=True)`
`self.add_constraint("cruise.aero_post.CMZ", lower=-0.01, upper=0.0, scaler=1.0)`
`self.add_constraint("geometry.linearcon", equals=0.0)`
[runScript_LoD.txt](https://github.com/mdolab/dafoam/files/11011202/runScript_LoD.txt)
[genFFD.txt](https://github.com/mdolab/dafoam/files/11011205/genFFD.txt)
[genAirFoilMesh.txt](https://github.com/mdolab/dafoam/files/11011208/genAirFoilMesh.txt)
[HS520PS.txt](https://github.com/mdolab/dafoam/files/11011211/HS520PS.txt)
[HS520SS.txt](https://github.com/mdolab/dafoam/files/11011212/HS520SS.txt)

**Comment by friedenhe** [comment]
If the baseline design's primal does not converge well, I suggest you change the mesh or BC to make it converge at least 5 orders. Adjoint-based optimization needs tight primal convergence. Otherwise, the adjoint gradient will be inaccurate and mislead the optimizer.
